google/resource_dataproc_cluster.go

```diff
@@ -0,0 +1,936 @@
+	if region == "global" && !zok {
+		return errors.New("zone is mandatory when region is set to 'global'")
+	}
+
+	// Create the cluster
+	op, err := config.clientDataproc.Projects.Regions.Clusters.Create(
+		project, region, cluster).Do()
+	if err != nil {
+		return err
+	}
+
+	d.SetId(clusterName)
+
+	// Wait until it's created
+	timeoutInMinutes := int(d.Timeout(schema.TimeoutCreate).Minutes())
+	waitErr := dataprocClusterOperationWait(config, op, "creating Dataproc cluster", timeoutInMinutes, 3)
+	if waitErr != nil {
+		// The resource didn't actually create
+		d.SetId("")
+		return waitErr
+	}
+
+	log.Printf("[INFO] Dataproc cluster %s has been created", clusterName)
+	return resourceDataprocClusterRead(d, meta)
+
+}
+
+func preemptibleInstanceGroupConfigCreate(cfg map[string]interface{}) *dataproc.InstanceGroupConfig {
+	icg := &dataproc.InstanceGroupConfig{}
+
+	if v, ok := cfg["num_instances"]; ok {
+		icg.NumInstances = int64(v.(int))
+	}
+	if dc, ok := cfg["disk_config"]; ok {
+		d := dc.([]interface{})
+		if len(d) > 0 {
+			dcfg := d[0].(map[string]interface{})
+			icg.DiskConfig = &dataproc.DiskConfig{}
+
+			if v, ok := dcfg["boot_disk_size_gb"]; ok {
+				icg.DiskConfig.BootDiskSizeGb = int64(v.(int))
+			}
+		}
+	}
+	return icg
+}
+
+func instanceGroupConfigCreate(cfg map[string]interface{}) *dataproc.InstanceGroupConfig {
+	icg := &dataproc.InstanceGroupConfig{}
+
+	if v, ok := cfg["num_instances"]; ok {
+		icg.NumInstances = int64(v.(int))
+	}
+	if v, ok := cfg["machine_type"]; ok {
+		icg.MachineTypeUri = extractLastResourceFromUri(v.(string))
+	}
+
+	if dc, ok := cfg["disk_config"]; ok {
+		d := dc.([]interface{})
+		if len(d) > 0 {
+			dcfg := d[0].(map[string]interface{})
+			icg.DiskConfig = &dataproc.DiskConfig{}
+
+			if v, ok := dcfg["boot_disk_size_gb"]; ok {
+				icg.DiskConfig.BootDiskSizeGb = int64(v.(int))
+			}
+			if v, ok := dcfg["num_local_ssds"]; ok {
+				icg.DiskConfig.NumLocalSsds = int64(v.(int))
+			}
+		}
+	}
+	return icg
+}
+
+func resourceDataprocClusterUpdate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	region := d.Get("region").(string)
+	clusterName := d.Get("name").(string)
+	timeoutInMinutes := int(d.Timeout(schema.TimeoutUpdate).Minutes())
+
+	cluster := &dataproc.Cluster{
+		ClusterName: clusterName,
+		ProjectId:   project,
+		Config:      &dataproc.ClusterConfig{},
+	}
+
+	updMask := []string{}
+
+	if d.HasChange("labels") {
+		v := d.Get("labels")
+		m := make(map[string]string)
+		for k, val := range v.(map[string]interface{}) {
+			m[k] = val.(string)
+		}
+		cluster.Labels = m
+
+		updMask = append(updMask, "labels")
+	}
+
+	if d.HasChange("cluster_config.0.worker_config.0.num_instances") {
+		wconfigs := d.Get("cluster_config.0.worker_config").([]interface{})
+		conf := wconfigs[0].(map[string]interface{})
+
+		desiredNumWorks := conf["num_instances"].(int)
```

Comment:
I think you can shorten these few lines just by doing `desiredNumWorks := d.Get("cluster_config.0.worker_config.0.num_instances").(int)`